src/Microsoft.Extensions.OData.Migration/UriSegmentTranslator.cs

```diff
@@ -0,0 +1,243 @@
+            return v4segment;
+        }
+
+        /// <summary>
+        /// Translates a V3 KeySegment (e.g. "/Boxes(1)") to a V4 KeySegment
+        /// </summary>
+        /// <param name="segment">OData V3 KeySegment</param>
+        /// <returns>OData V4 KeySegment</returns>
+        public override ODataPathSegment Translate(Data.OData.Query.SemanticAst.KeySegment segment)
+        {
+            IEdmNavigationSource v4navigationSource = v4model.FindDeclaredNavigationSource(segment.EntitySet.Name);
+            IEdmEntityType v4type = (IEdmEntityType)v4model.FindType(segment.EdmType.FullTypeName());
+            KeySegment v4segment = new KeySegment(segment.Keys, v4type, v4navigationSource);
+            return v4segment;
+        }
+
+        /// <summary>
+        /// Translates at V3 PropertySegment (e.g. "/Boxes(1)/Name") into a V4 PropertySegment
+        /// </summary>
+        /// <param name="segment">OData V3 PropertySegment</param>
+        /// <returns>OData V4 PropertySegment</returns>
+        public override ODataPathSegment Translate(Data.OData.Query.SemanticAst.PropertySegment segment)
+        {
+            // Extract the DeclaringType (e.g. EntitySet) name from V3 PropertySegment
+            string v3typeName = segment.Property.DeclaringType.FullTypeName();
+
+            // Look up V4 Structured Type (guaranteed if found) in V4 model
+            IEdmStructuredType v4Type = v4model.FindType(v3typeName) as IEdmStructuredType;
+
+            // Look up corresponding Property in V4 Type by name
+            IEdmStructuralProperty v4Property = v4Type.FindProperty(segment.Property.Name) as IEdmStructuralProperty;
+
+            return new PropertySegment(v4Property);
+        }
+
+        /// <summary>
+        /// Translates a V3 OperationSegment to V4 OperationSegment by locating all matching function imports in the v4 model
+        /// and constructing a new V4 OperationSegment with all found operation segments.
+        /// </summary>
+        /// <param name="segment"></param>
+        /// <returns></returns>
+        public override ODataPathSegment Translate(Data.OData.Query.SemanticAst.OperationSegment segment)
+        {
+            List<IEdmOperationImport> v4OpImports = new List<IEdmOperationImport>();
+            List<IEdmOperation> v4UnboundOps = new List<IEdmOperation>();
+            List<IEdmOperation> v4BoundOps = new List<IEdmOperation>();
+
+            IEdmEntitySet v4entitySet = segment.EntitySet != null ? v4model.FindDeclaredEntitySet(segment.EntitySet.Name) : null;
+
+            // For each operation in V3 operation segment, determine corresponding action/function/function imports in v4
+            foreach (Data.Edm.IEdmFunctionImport op in segment.Operations)
+            {
+                IEnumerable<Data.Edm.IEdmFunctionParameter> entityParams = op.Parameters.Where(p => p.Type.Definition.GetType().GetInterfaces().Select(iface => iface.Name).Contains("IEdmEntityType"));

```

Comment:
Select(iface => iface.Name).Contains("IEdmEntityType") be simplified to Any(i => string.equal(i.Name, "IEdmEntityType")